File: java/bob/src/main/java/Bob.java
class Bob {

  String hey(String input) {

    if (input.isBlank()) {
      return "Fine. Be that way!";
    } else if (atLeastOneAlpha(input) && input.toUpperCase().equals(input)
        && input.charAt(input.length() - 1) == '?') {
      return "Calm down, I know what I'm doing!";
    } else if (input.trim().charAt(input.trim().length() - 1) == '?') {
      return "Sure.";
    } else if (atLeastOneAlpha(input) && input.toUpperCase().equals(input)) {
      return "Whoa, chill out!";
    }

    return "Whatever.";
  }

  boolean atLeastOneAlpha(String input) {
    return input.matches(".*[a-zA-Z]+.*");
  }
}

File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1]).

response(_String) ->
    Cleaned = string:trim(_String),
    case Cleaned of
        "" -> "Fine. Be that way!";
        _ -> message_filter(Cleaned)
    end.

message_filter(Message) ->
    Letters = lists:filter(fun(C) -> is_letter(C) end, Message),
    IsQuestion = lists:last(Message) == $?,
    AllUpperCase = all_uppercase(Letters),
    case {IsQuestion, AllUpperCase, Letters =/= []} of
        {true, true, true} -> "Calm down, I know what I'm doing!";
        {true, _, _} -> "Sure.";
        {_, true, true} -> "Whoa, chill out!";
        _ -> "Whatever."
    end.

is_letter(C) ->
    C >= $A andalso C =< $Z orelse C >= $a andalso C =< $z.

all_uppercase(Letters) ->
    lists:all(fun(C) -> C >= $A andalso C =< $Z end, Letters).
